sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -852,23 +835,171 @@ case class GpuRegExpReplace(
+          val cudfRegexPattern = new CudfRegexTranspiler(replace = false)
+            .transpile(javaRegexpPattern)
+          pattern = Some(cudfRegexPattern)
+          numGroups = countGroups(new RegexParser(javaRegexpPattern).parse())
+        } catch {
+          case e: RegexUnsupportedException =>
+            willNotWorkOnGpu(e.getMessage)
+        }
+      case _ =>
+        willNotWorkOnGpu(s"only non-null literal strings are supported on GPU")
+    }
+
+    val idx = expr.idx.asInstanceOf[Literal].value.asInstanceOf[Int]
+    if (idx < 0) {
+      throw new IllegalArgumentException("The specified group index cannot be less than zero")
+    }
+    if (idx > numGroups) {
+      throw new IllegalArgumentException(
+        s"Regex group count is $numGroups, but the specified group index is $idx")
+    }
+  }
+
+  override def convertToGpu(
+      str: Expression,
+      regexp: Expression,
+      idx: Expression): GpuExpression = {
+    val cudfPattern = pattern.getOrElse(
+      throw new IllegalStateException("Expression has not been tagged with cuDF regex pattern"))
+    GpuRegExpExtract(str, regexp, idx, cudfPattern)
+  }
+}
+
+case class GpuRegExpExtract(
+    subject: Expression,
+    regexp: Expression,
+    idx: Expression,
+    cudfRegexPattern: String)
+  extends GpuRegExpTernaryBase with ImplicitCastInputTypes with NullIntolerant {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType, IntegerType)
+  override def first: Expression = subject
+  override def second: Expression = regexp
+  override def third: Expression = idx
+
+  override def dataType: DataType = StringType
+  override def prettyName: String = "regexp_extract"
+
+  override def doColumnar(
+      str: GpuColumnVector,
+      regexp: GpuScalar,
+      idx: GpuScalar): ColumnVector = {
+
+    val groupIndex = idx.getValue.asInstanceOf[Int]
+
+    // There are some differences in behavior between cuDF and Java so we have
+    // to handle those cases here.
+    //
+    // Given the pattern `^([a-z]*)([0-9]*)([a-z]*)$` the following table
+    // shows the value that would be extracted for group index 2 given a range
+    // of inputs. The behavior is mostly consistent except for the case where
+    // the input is non-null and does not match the pattern.
+    //
+    // | Input  | Java  | cuDF  |
+    // |--------|-------|-------|
+    // | ''     | ''    | ''    |
+    // | NULL   | NULL  | NULL  |
+    // | 'a1a'  | '1'   | '1'   |
+    // | '1a1'  | ''    | NULL  |
+
+    if (groupIndex == 0) {
+      withResource(GpuScalar.from("", DataTypes.StringType)) { emptyString =>
+        withResource(GpuScalar.from(null, DataTypes.StringType)) { nullString =>
+          withResource(str.getBase.matchesRe(cudfRegexPattern)) { matches =>
+            withResource(str.getBase.isNull) { isNull =>
+              withResource(matches.ifElse(str.getBase.incRefCount(), emptyString)) {
```

Comment:
Why the `str.getBase.incRefCount()`?  It seems that would trigger a refcount-induced leak.  Applies to same code below.

Reply:
This is one of the reasons I moved the PR back to draft. I did see a leak yesterday during testing but removing these `incrRefCount` calls results in a `closed too many times` error, so I need to figure out what is going on here.

Reply:
I think the issue has to do with this:
```
          withResource(str.getBase.extractRe(cudfRegexPattern)) { extract =>
[...]
                withResource(extract.getColumn(groupIndex - 1)) { extractedGroup =>
```
`extractRe` returns a `Table` which we definitely need to close, but `Table.getColumn` does not increment ref counts, so we should not be putting that in a `withResource` and trying to close it additionally.  Drop the `withResource` on the `getColumn` line and remove the extraneous `incRefCount` stuff and it should be better.